crates/fuel-core/src/query/message.rs

```diff
@@ -167,52 +152,107 @@ pub fn message_proof<T: MessageProofData + ?Sized>(
-                data: message_data,
+                data,
                 ..
             } if r.message_id() == Some(message_id) => {
-                Some((sender, recipient, nonce, amount, message_data))
+                Some((sender, recipient, nonce, amount, data))
             }
             _ => None,
         });
 
-    let (sender, recipient, nonce, amount, message_data) = match receipt {
+    let (sender, recipient, nonce, amount, data) = match receipt {
         Some(r) => r,
         None => return Ok(None),
     };
 
     // Get the block id from the transaction status if it's ready.
-    let block_id = match data
+    let message_block_id = match database
         .transaction_status(&transaction_id)
         .into_api_result::<TransactionStatus, StorageError>()?
     {
         Some(TransactionStatus::Success { block_id, .. }) => block_id,
         _ => return Ok(None),
     };
 
-    // Get the message ids in the same order as the transactions.
-    let leaves: Vec<Vec<Receipt>> = data
-        .transactions_on_block(&block_id)?
-        .into_iter()
-        .map(|id| data.receipts(&id))
+    // Get the message fuel block header.
+    let (message_block_header, message_block_txs) = match database
+        .block(&message_block_id)
+        .into_api_result::<CompressedBlock, StorageError>()?
+    {
+        Some(t) => t.into_inner(),
+        None => return Ok(None),
+    };
+
+    let (root, message_proof) =
+        match message_receipts_proof(database, message_id, &message_block_txs)? {
+            Some(proof) => proof,
+            None => return Ok(None),
+        };
+
+    if message_block_header.message_receipt_root != root {
+        // This is bad as it means there's a bug in our prove code or block production.
+        Err(anyhow::anyhow!(
+            "block header {:?} root doesn't match generated proof root {:?}",
+            message_block_header,
+            root
+        ))?;
+    }
+
+    // Get the commit fuel block header.
+    let commit_block_header = match database
+        .block(&commit_block_id)
+        .into_api_result::<CompressedBlock, StorageError>()?
+    {
+        Some(t) => t.into_inner().0,
+        None => return Ok(None),
+    };
+
+    let verifiable_commit_block_height = *commit_block_header.height() - 1u32.into();
+    let block_proof = database.block_history_proof(
+        message_block_header.height(),
+        &verifiable_commit_block_height,
+    )?;
+
+    Ok(Some(MessageProof {
+        message_proof,
+        block_proof,
+        message_block_header,
+        commit_block_header,
+        sender,
+        recipient,
+        nonce,
+        amount,
+        data,
+    }))
+}
+
+fn message_receipts_proof<T: MessageProofData + ?Sized>(
+    database: &T,
+    message_id: MessageId,
+    message_block_txs: &[Bytes32],
+) -> StorageResult<Option<(Bytes32, MerkleProof)>> {
+    // Get the message receipts from the block.
+    let leaves: Vec<Vec<Receipt>> = message_block_txs
```

Comment:
I'm not very familiar with how receipts are created, but is it possible to have storage-backed Merkle trees for receipts, and append to the Merkle tree whenever we create new receipts? It looks like we might be able to use per-message BMTs (keyed by `message_id`)?

Reply:
Usually, we don't have too many messages in the block, so it is faster to calculate that in runtime than have access to the storage, I guess. 

But if we have a lot of messages, yeah, optimization like this would not be bad. Bad it's more about GraphQL performance, and maybe should be solved on the indexer side